cpp/src/arrow/array_list_test.cc

```diff
@@ -523,6 +526,161 @@ TEST_F(TestMapArray, BuildingStringToInt) {
+                     offsets3->data()->buffers[0], 1);
+  AssertArraysEqual(expected3, *map3);
+
+  // Check that the last offset bit is zero
+  ASSERT_FALSE(BitUtil::GetBit(map3->null_bitmap()->data(), length + 1));
+
+  MapArray expected4(map_type, length, offsets2->data()->buffers[1], keys, items,
+                     offsets4->data()->buffers[0], 1);
+  AssertArraysEqual(expected4, *map4);
+
+  // Test failure modes
+
+  std::shared_ptr<Array> tmp;
+
+  // Zero-length offsets
+  ASSERT_RAISES(Invalid,
+                MapArray::FromArrays(offsets1->Slice(0, 0), keys, items, pool_, &tmp));
+
+  // Offsets not the right type
+  ASSERT_RAISES(TypeError, MapArray::FromArrays(keys, offsets1, items, pool_, &tmp));
+
+  // Keys and Items different lengths
+  ASSERT_RAISES(Invalid,
+                MapArray::FromArrays(offsets1, keys->Slice(0, 1), items, pool_, &tmp));
+
+  // Keys contains null values
+  std::shared_ptr<Array> keys_with_null = offsets3;
+  std::shared_ptr<Array> tmp_items = items->Slice(0, offsets3->length());
+  ASSERT_EQ(keys_with_null->length(), tmp_items->length());
+  ASSERT_RAISES(Invalid,
+                MapArray::FromArrays(offsets1, keys_with_null, tmp_items, pool_, &tmp));
+}
+
+namespace {
+
+template <typename TYPE>
+Status BuildListOfStructPairs(TYPE& builder, std::shared_ptr<Array>* out) {
+  auto struct_builder = internal::checked_cast<StructBuilder*>(builder.value_builder());
+  auto field0_builder =
+      internal::checked_cast<Int16Builder*>(struct_builder->field_builder(0));
+  auto field1_builder =
+      internal::checked_cast<Int16Builder*>(struct_builder->field_builder(1));
+
+  RETURN_NOT_OK(builder.Append());
+  RETURN_NOT_OK(field0_builder->AppendValues({0, 1}));
+  RETURN_NOT_OK(field1_builder->AppendValues({1, -1}, {1, 0}));
+  RETURN_NOT_OK(struct_builder->AppendValues(2, NULLPTR));
+  RETURN_NOT_OK(builder.AppendNull());
+  RETURN_NOT_OK(builder.Append());
+  RETURN_NOT_OK(field0_builder->Append(2));
+  RETURN_NOT_OK(field1_builder->Append(3));
+  RETURN_NOT_OK(struct_builder->Append());
+  RETURN_NOT_OK(builder.Append());
+  RETURN_NOT_OK(builder.Append());
+  RETURN_NOT_OK(field0_builder->AppendValues({3, 4}));
+  RETURN_NOT_OK(field1_builder->AppendValues({4, 5}));
+  RETURN_NOT_OK(struct_builder->AppendValues(2, NULLPTR));
+  RETURN_NOT_OK(builder.Finish(out));
+  RETURN_NOT_OK((*out)->Validate());
+
+  return Status::OK();
+}
+
+}  // namespace
+
+TEST_F(TestMapArray, ValueBuilder) {
+  auto key_builder = std::make_shared<Int16Builder>();
+  auto item_builder = std::make_shared<Int16Builder>();
+  MapBuilder map_builder(default_memory_pool(), key_builder, item_builder);
+
+  // Build Map array using key/item builder
+  std::shared_ptr<Array> expected;
+  ASSERT_OK(map_builder.Append());
+  ASSERT_OK(key_builder->AppendValues({0, 1}));
+  ASSERT_OK(item_builder->AppendValues({1, -1}, {1, 0}));
+  ASSERT_OK(map_builder.AppendNull());
```

Comment:
In this case `AdjustStructBuilderLength` needs to be called after `key_builder` and `item_builder` have appended items to fix the `struct_builder` length, so it is called at the beginning of `map_builder.Append***` before starting the next slot.